edb/edgeql/compiler/viewgen.py

```diff
@@ -98,6 +98,21 @@ class ShapePtr(NamedTuple):
     target_set: Optional[irast.Set]
 
 
+@dataclasses.dataclass(kw_only=True, frozen=True)
+class ShapeContext:
+    # a helper object for passing shape compile parameters
+
+    ctx: context.ContextLevel
```

Comment:
Hm... Not 100% sold on including ctx, but I dont' feel strongly

Reply:
Yeah, I think I'd prefer it to remain an explicit argument.